protocol_grpc.go

```diff
@@ -750,20 +768,34 @@ func grpcHTTPToCode(httpCode int) Code {
 // binary Protobuf format, even if the messages in the request/response stream
 // use a different codec. Consequently, this function needs a Protobuf codec to
 // unmarshal error information in the headers.
-func grpcErrorFromTrailer(bufferPool *bufferPool, protobuf Codec, trailer http.Header) *Error {
+func grpcErrorFromHeaders(bufferPool *bufferPool, protobuf Codec, trailer, header http.Header) *Error {
 	codeHeader := trailer.Get(grpcHeaderStatus)
-	if codeHeader == "" || codeHeader == "0" {
+	if codeHeader == "" {
+		// Can send trailers with response headers if there is no message.
+		codeHeader = header.Get(grpcHeaderStatus)
+		if codeHeader == "" {
+			return errTrailersWithoutGRPCStatus
+		}
+	}
+	if codeHeader == "0" {
 		return nil
 	}
 
 	code, err := strconv.ParseUint(codeHeader, 10 /* base */, 32 /* bitsize */)
 	if err != nil {
 		return errorf(CodeInternal, "gRPC protocol error: invalid error code %q", codeHeader)
 	}
-	message := grpcPercentDecode(bufferPool, trailer.Get(grpcHeaderMessage))
+	messageHeader := trailer.Get(grpcHeaderMessage)
+	if messageHeader == "" {
+		messageHeader = header.Get(grpcHeaderMessage)
```

Comment:
This seems to be saying that if you can't get something from the trailer, then get it from the header as a fallback - is this part of the gRPC protocol specification?

Reply:
Changed the logic to only get values from the trailers or headers (depending on where we find the `grpc-status` header), but not from both. I believe the relevant part of the spec is here: https://github.com/grpc/grpc/blob/master/doc/PROTOCOL-HTTP2.md#responses.

> Response → (Response-Headers *Length-Prefixed-Message Trailers) / _Trailers-Only_

I'll walk through the failing gRPC-Web test again to make sure we're doing things right on our end and also run the crosstests against this branch.

Reply:
`Status must be sent in Trailers even if the status code is OK.`

This would seem to imply that the status can only be sent in trailers, right?

Reply:
A Trailers-Only response returns HTTP-Status, Content-Type, and Trailers (Status [Status-Message] *Custom-Metadata). For gRPC-Web, we match the behavior in Envoy to skip encoding trailers in the message body and just return them as regular headers: https://github.com/bufbuild/connect-go/blob/871250f0e169d8afe8fe097165a025e1a558e1f8/protocol_grpc.go#L494-L504

I've taken a conservative approach with this PR and only look in both trailers and then headers for the gRPC-Web client. If we see in the future that we need to support a Trailing-Headers response for the gRPC protocol as well, we'd just update https://github.com/bufbuild/connect-go/pull/288 to pass `r.header` instead of `nil`.

Reply:
The protocol's logical trailers can be sent as HTTP trailers, or in some cases, as HTTP headers. Confusingly, the latter case is called a "trailers-only" response.